README.md
# :mag_right: CSV-Compare
[![PyPI version](https://badge.fury.io/py/csv-compare-tool.svg)](https://badge.fury.io/py/csv-compare-tool)
[![Build Status](https://github.com/liquidz00/csv-compare/actions/workflows/test.yaml/badge.svg)](https://github.com/liquidz00/csv-compare/actions/workflows/test.yaml)
#### A python package for Jamf MacAdmins in mind

The csvcomparetool allows you to find differences between two CSV files based on a specified column identifier. It provides a simple way to compare the contents of two CSV files and identify which records are present in one file but not in the other.

> [!NOTE]
> When comparing CSV files for differences, be sure to provide the CSV with more entries second. For example, if CSV-1 has a list of 34 names, and CSV-2 has a list of 40, CSV-2 should be set as the second passed CSV path in order for differences to show as expected.

- [Installation](#installation)
- [Usage](#usage)
  - [Import class](#import-the-csvcomparer-class-from-the-package)
  - [Create object](#create-a-csvcomparer-object)
  - [Validate paths](#validate-the-paths-to-ensure-the-specified-csv-files-exist)
  - [Validate column](#validate-the-column-exists-in-the-provided-csv-files)
  - [Find differences](#find-the-differences-between-the-two-csv-files)
  - [Handle differences as needed](#print-or-process-the-differences-as-needed)
- [Examples](#examples)
  - [Python project](#python-project)
  - [CLI](#command-line-tool)

## Installation

You can install CSV-Compare using `pip`:
```bash
pip install csv-compare-tool
```

## Usage
To use CSV-Compare in your Python project, follow these steps.
> [!IMPORTANT]
> When importing as a package in a Python project, be sure to **remove** the hyphens so the package reads `csvcomparetool` (See [PEP 8](https://peps.python.org/pep-0008/#package-and-module-names) for more information)

### Import the package
```python
import csvcomparetool
```
Optionally, you can import the `CSVComparer` class alone
```python
from csvcomparetool import CSVComparer
```

### Create a `CSVComparer` object
```python
csv1_path = "path/to/first.csv"
csv2_path = "path/to/second.csv"
column = "identifier_column"
comparer = CSVComparer(csv1_path, csv2_path, column)
```

### Validate the paths to ensure the specified CSV files exist
```python
if not comparer.validate_paths():
    print("CSV file paths are invalid. Please check the file paths and try again.")
    return
```

### Validate the column exists in the provided CSV files
```python
if not comparer.validate_columns():
    print("Provided column not found in CSV. Check the columns and try again.")
    return
```

### Find the differences between the two CSV files
```python
differences = comparer.find_differences()
```

### Print or process the differences as needed
```python
for difference in differences:
    print(f"Record '{difference}' is present in CSV2 but not in CSV1.")
```

## Examples
### Python Project

```python
from csvcomparetool import CSVComparer

csv1_path = "path/to/first.csv"
csv2_path = "path/to/second.csv"
column = "identifier_column"

comparer = CSVComparer(csv1_path, csv2_path, column)

if not comparer.validate_paths() or not comparer.validate_columns():
    print("CSV file paths are invalid, or the column identifier does not exist. Check the file paths and columns and try again.")
else:
    differences = comparer.find_differences()
    for difference in differences:
        print(f"Record '{difference}' is present in CSV2 but not in CSV1.")
```

### Command Line tool
First, clone this repository to a local directory on your machine
```bash
git clone https://github.com/liquidz00/csv-compare.git
```
Navigate to the cloned repo location (local directory where you chose to save the repository)
```bash
cd /path/to/repo/src/csvcomparetool/
```
Lastly, run the following command
```bash
python cli.py /csv/path/one /csv/path/two columnidentifier
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. 

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Features currently in progress:

- Adding additional column identifier in event columns are not the same across the two CSV files

## [1.0.3] - 2023-09-15

### Fixed

- Issue with project files not properly being indexed in distribution being uploaded to PyPi
- Issue where `CSVComparer` class could not be imported after installing via `pip`

### Deprecated

- Yanked version 1.0.2 from PyPi due to package import issues mentioned above

## [1.0.2] - 2023-09-13

### Changed

- Configuration of pyproject.toml file
- Refactored package name from csv-compare-tool to csvcomparetool for easier import statements
- Bumped to v1.0.2

### Deprecated

- `setup.py` file per [setuptools documentation](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#transitioning-from-setup-py-to-setup-cfg)

## [1.0.1] - 2023-09-12

### Fixed

- Issue with build when installing package via `pip`

### Changed

- Bumped to version 1.0.1
- Metadata in pyproject.toml and setup.py files




tests/test_csv_comparer.py
import pytest
import os

from src.csvcomparetool.csvcomparetool import CSVComparer

BASE_DIR = os.path.dirname(__file__)

@pytest.fixture
def fruit_csvs():
    fruits1 = os.path.join(BASE_DIR, "fruits1.csv")
    fruits2 = os.path.join(BASE_DIR, "fruits2.csv")

    yield fruits1, fruits2

@pytest.fixture
def title_csvs():
    titles1 = os.path.join(BASE_DIR, "anyorg_old_titles.csv")
    titles2 = os.path.join(BASE_DIR, "anyorg_new_titles.csv")

    yield titles1, titles2

def test_fruits(fruit_csvs):
    csv1, csv2 = fruit_csvs
    col = "Fruit"

    comparer = CSVComparer(csv1, csv2, col).find_differences()

    expected_result = ["Banana"]

    assert comparer == expected_result

def test_titles(title_csvs):
    csv1, csv2 = title_csvs
    col = "Job Title"

    comparer = CSVComparer(csv1, csv2, col).find_differences()

    expected_result = ['Product Designer', 'Data Analyst', 'Software Developer', 'HR Administrator', 'Market Researcher', 'Quality Assurance Specialist', 'Support Specialist', 'Recruitment Manager', 'Public Relations Specialist', 'IT Support Specialist', 'Market Analyst', 'Customer Relations Specialist']

    assert comparer == expected_result

def test_invalid_column(title_csvs):
    with pytest.raises(Exception) as exc_info:
        csv1, csv2 = title_csvs
        CSVComparer(csv1, csv2, identifier="invalid column").find_differences()
        raise Exception("ERROR: Column identifier does not exist. Try again.")

    assert exc_info.value.args[0] == "ERROR: Column identifier does not exist. Try again."






csv_compare/CSVComparer.py
# Author: liquidz00
# GitHub: https://github.com/liquidz00
#
#
# Written: Sep 08 23

import pandas as pd
import logging
import os
from typing import Optional, Union

# Logger
logger = logging.getLogger(__name__)

# StrPath
StrPath = Union[str, 'os.PathLike[str]']

class CSVComparer:
    def __init__(self, path_one: Optional[StrPath], path_two: Optional[StrPath], identifier: Optional[str]) -> None:
        self.path_one: Optional[StrPath] = path_one
        self.path_two: Optional[StrPath] = path_two
        self.identifier: Optional[str] = identifier

    def validate_paths(self):
        if not os.path.exists(self.path_one):
            logging.error(f"CSV file not found: {self.path_one}")
            return False
        if not os.path.exists(self.path_two):
            logging.error(f"CSV file not found: {self.path_two}")
            return False
        return True

    def validate_columns(self):
        db_1 = pd.read_csv(self.path_one)
        db_2 = pd.read_csv(self.path_two)

        if self.identifier not in db_1.columns:
            logging.error(f"Column '{self.identifier}' not found in {self.path_one}")
            return False
        if self.identifier not in db_2.columns:
            logging.error(f"Column '{self.identifier}' not found in {self.path_two}")
            return False
        return True

    def find_differences(self):
        if not self.validate_paths() or not self.validate_columns():
            return []

        db_1 = pd.read_csv(self.path_one)
        db_2 = pd.read_csv(self.path_two)

        users_1 = db_1[self.identifier].to_list()
        users_2 = db_2[self.identifier].to_list()

        differences = []

        for user in users_2:
            if user not in users_1:
                differences.append(user)

        return differences


src/csvcomparetool/cli.py
# Author: liquidz00
# GitHub: https://github.com/liquidz00
#
#
# Written: Sep 08 23

import argparse
from src.csvcomparetool.csvcomparetool import CSVComparer

help_desc = """

CSV Compare Tool // github.com/liquidz00 // AUG 2023

Compare two different CSV files by specifying a column identifier. 

Usage (after navigating to project directory in Terminal): 
python3 cli.py path/to/first.csv path/to/second.csv column

Where 'column' is the column title containing data to compare

"""

def main():
    parser = argparse.ArgumentParser(description=help_desc, formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('csv1', '-c1', type=str, help='Path to the first CSV file')
    parser.add_argument('csv2', '-c2', type=str, help='Path to the second CSV file')
    parser.add_argument('column', '-column', type=str, help='Column identifier for both CSVs')

    args = parser.parse_args()

    comparer = CSVComparer(args.csv1, args.csv2, args.column)

    if not comparer.validate_paths():
        print("CSV file paths are invalid. Please check the file paths and try again.")
        return

    if not comparer.validate_columns():
        print("Invalid column. Please check column identifier and try again.")

    differences = comparer.find_differences()

    if differences:
        print("Differences found:")
        for difference in differences:
            print(f"Record '{difference}' is present in CSV2 but not in CSV1.")
    else:
        print("No differences found.")

if __name__ == "__main__":
    main()
